autopush/tests/test_integration.py

```diff
@@ -1936,6 +1939,253 @@ def test_registration(self):
+        )
+        print ("Response: %s" % response.code)
+        assert response.code == 201
+
+        ca_data = self._mock_send.post.mock_calls[1][2]['json']['data']
+        # ChannelID here MUST match what we got from the registration call.
+        # Currently, this is a lowercase, hex UUID without dashes.
+        assert ca_data['chid'] == jbody['channelID']
+        assert ca_data['con'] == content_encoding
+        assert ca_data['cryptokey'] == crypto_key
+        assert ca_data['enc'] == salt
+        assert ca_data['body'] == base64url_encode(data)
+
+    @inlineCallbacks
+    def test_bad_registration(self):
+        self._add_router()
+        url = "{}/v1/{}/{}/registration".format(
+            self.ep.conf.endpoint_url,
+            "adm",
+            "dev",
+        )
+        self._mock_reply.status_code = 400
+        response, body = yield _agent("POST", url, body=json.dumps(
+            {"token": self.token[:-100]}
+        ))
+        assert response.code == 400
+
+    @inlineCallbacks
+    def test_bad_token_refresh(self):
+        self._add_router()
+        url = "{}/v1/{}/{}/registration".format(
+            self.ep.conf.endpoint_url,
+            "adm",
+            "dev",
+        )
+        self._mock_reply.status_code = 200
+        response, body = yield _agent("POST", url, body=json.dumps(
+            {"token": self.token}
+        ))
+        assert response.code == 200
+        jbody = json.loads(body)
+        data = ("\xa2\xa5\xbd\xda\x40\xdc\xd1\xa5\xf9\x6a\x60\xa8\x57\x7b\x48"
+                "\xe4\x43\x02\x5a\x72\xe0\x64\x69\xcd\x29\x6f\x65\x44\x53\x78"
+                "\xe1\xd9\xf6\x46\x26\xce\x69")
+        crypto_key = ("keyid=p256dh;dh=BAFJxCIaaWyb4JSkZopERL9MjXBeh3WdBxew"
+                      "SYP0cZWNMJaT7YNaJUiSqBuGUxfRj-9vpTPz5ANmUYq3-u-HWOI")
+        salt = "keyid=p256dh;salt=S82AseB7pAVBJ2143qtM3A"
+        content_encoding = "aesgcm"
+        self._mock_reply.status_code = 400
+        self._mock_reply.text = "Test error"
+        self._mock_reply.content = "Test content"
+        response, body = yield _agent(
+            'POST',
+            str(jbody['endpoint']),
+            headers=Headers({
+                "crypto-key": [crypto_key],
+                "encryption": [salt],
+                "ttl": ["0"],
+                "content-encoding": [content_encoding],
+            }),
+            body=data
+        )
+        assert response.code == 500
+
+    def test_bad_sends(self):
+        from requests.exceptions import ConnectionError
+        self._add_router()
+
+        # prevent the token refresh check
+        self.ep.routers["adm"].profiles["dev"]._auth_token = "SomeToken"
+        self.ep.routers["adm"].profiles["dev"]._token_exp = time.time() + 300
+        self._mock_reply.status_code = 400
+        self._mock_reply.text = "Test error"
+        self._mock_reply.content = "Test content"
+
+        # This will call RouterException, which may halt the test thread.
```

Comment:
I don't get this comment, why would RouterException halt the test thread?

Reply:
Since we'd be calling via the `_agent`, `RouterException` is passed up and not trapped, causing `_agent` to fail out and kill/fail the test. I was not able to trap for it. Thus calling the send directly with the raise trap. I can try to muck about more with `_agent` to try and handle that sort of thing, but seems a bit out of scope for this patch.

Reply:
That might be a case of twisted's TestCase failing if it sees a log.failure call made during the test. You can call self.flushLoggedErrors() at the end of tests to stop it from doing that

Reply:
@pjenvey Ah, cool. I'll try that.